CMake cleanup

pfultz2 opened the pull request:
This removes the object libraries and uses normal cmake targets. This helps simplifies a lot of the cmake since we can properly propagate usage requirements.

pfultz2 commented:
Because of the global registration an object library will need to be used for at least cppcheck-core, but we can still wrap it in an interface library so it can propagate usage requirements though.

firewave commented:
Using interface libraries would be an acceptable modernization (although I personally detest them - I had to many problems with them in the past). We need to test though that things work with the currently specified minimum CMake version and raise it appropriately if necessary.

But if we actually want to support shared libraries there is *a lot* of work to be done:
- visibility handling on non-Windows platforms
- needs to add annotation to all exported symbols (currently it was only set for the ones actually used in tests)
- needs ABI handling/versioning for the shared objects
- adjust the install targets
- adjust the shipped Visual Studio project to generate the same output files
- adjust the Windows installer
- CI jobs to test this on all platforms
- ... possibly even more things I cannot think off the top of my head

And IMO externals should not be shared libraries.

Also we still have static objects in our code which possibly makes the proper usage of the shared objects impossible. Beside that I reckon there never was and never will be a single user of the shared objects. And the current implementation was only partial.

We should not be supporting shared libraries at all because it just causes more work now and in the future and I do not see *any* upsides at all. We should be removing features instead of adding and simplifying things instead of making them even more complex (with latter I am referring to the support matrix - not the code).

pfultz2 commented:
The purpose of the shared libraries would only be for local dev. I wouldn't envision us installing them.

firewave commented:
> The purpose of the shared libraries would only be for local dev. 

Please elaborate. I do not see how it would be of any use - especially given that it was never available and nobody complained. And if we want to support that we need to test it i.e. CI workflows. Otherwise it is just more stuff which rots away in this repo.

> I wouldn't envision us installing them.

It would be possible but it would not work - so it should not be allowed.

pfultz2 commented:
> Please elaborate. I do not see how it would be of any use

Faster linking time.

firewave commented on externals/tinyxml2/CMakeLists.txt:
```diff
@@ -20,5 +20,5 @@ if(CYGWIN)
 endif()
 
-target_include_directories(tinyxml2 SYSTEM PUBLIC .)
+target_include_directories(tinyxml2 PUBLIC .)
```
That flags needs to depend on the `EXTERNALS_AS_SYSTEM` option.

firewave commented on lib/CMakeLists.txt:
```diff
@@ -70,6 +70,6 @@ if (Boost_FOUND)
-if (NOT CMAKE_DISABLE_PRECOMPILE_HEADERS)
-    target_precompile_headers(cppcheck-core-private PRIVATE precompiled.h)
-endif()
+# if (NOT CMAKE_DISABLE_PRECOMPILE_HEADERS)
+#     target_precompile_headers(cppcheck-core-private PRIVATE precompiled.h)
+# endif()
```
Needs to be enabled again before merging.

pfultz2 replied:
Yea, that was to test the CI. I was getting error:

```
../../../lib/CMakeFiles/cppcheck-core-private.dir/cmake_pch.hxx.gch: file not recognized: file format not recognized
```

Now with it disabled I get an error that there isnt a make rule for it.

firewave commented on externals/tinyxml2/CMakeLists.txt:
```diff
@@ -1,24 +1,24 @@
-if (BUILD_CORE_DLL)
-    target_compile_definitions(tinyxml2_objs PRIVATE TINYXML2_EXPORT)
-endif()
+add_library(tinyxml2 ${srcs} ${hdrs})
```
This should be able to become a shared library.

firewave replied:
That should have said "should not be able".

pfultz2 commented on externals/simplecpp/CMakeLists.txt:
```diff
@@ -1,11 +1,11 @@
-if (BUILD_CORE_DLL)
-    target_compile_definitions(simplecpp_objs PRIVATE SIMPLECPP_EXPORT)
-endif()
+add_library(simplecpp ${srcs} ${hdrs})
```
It will if we set `BUILD_SHARED_LIBS`.

pfultz2 commented on frontend/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
-add_library(frontend_objs OBJECT ${hdrs} ${srcs})
-target_include_directories(frontend_objs PRIVATE ${PROJECT_SOURCE_DIR}/lib)
\ No newline at end of file
+add_library(frontend OBJECT ${hdrs} ${srcs})
```
I see why this works. Since cmake 3.12, object libraries can be linked directly instead of adding the `$<TARGET_OBJECTS>` to the sources. This will make it simpler.

firewave replied:
> I see why this works. Since cmake 3.12, object libraries can be linked directly instead of adding the `$<TARGET_OBJECTS>` to the sources. This will make it simpler.

Doesn't that mean that we do not actually need real libraries in CMake?

pfultz2 commented:
@firewave Alright, I think this is ready to go. This will fix the issue with using boost(#7655 is still needed if we want to disable boost).

firewave commented on cli/cppcheckexecutor.cpp:
```diff
@@ -734,7 +734,7 @@ int CppCheckExecutor::executeCommand(std::string exe, std::vector<std::string> a
     const int res = _pclose(p);
 #else
-    const int res = pclose(p);
+    int res = pclose(p);
```
The scope is getting way too big with a lot of questionable sidecar changes. This needs to be fixed more incrementally.

> Its needed to fix CI, because they became errors in the CI with these changes. I am not sure why it didnt error out before in the CI.

That should not be happening. That indicates that something is wrong in these CMake change.

firewave commented:
> @firewave Alright, I think this is ready to go. This will fix the issue with using boost(#7655 is still needed if we want to disable boost).

There's several things which do not look right to me. I am not feeling very well right now so it will take quite a while for me to review and even understand what this is doing since the changes are hard to follow.

Also the proper fixes for the boost issues should be merged as since those will have actual tests for the behavior changes.

pfultz2 commented:
> There's several things which do not look right to me. I am not feeling very well right now so it will take quite a while for me to review and even understand what this is doing since the changes are hard to follow.

Hope you feel better soon. Here is a summary/explanation of the changes which hopefully can help you review the PR.

The biggest change is to move all the usage requirements into the target and link with target directly instead of using `$<TARGET_OBJECTS>` and manually adding it for each consumer. So writing `target_link_libraries(someTarget cppcheck-core)` it will link with cppcheck-core and add any headers or defines that are need to use cppcheck-core.

So now I removed the manual includes and linking, and just link a single target directly. I did this based on the includes in the cmake, but I didnt check the source code if its actually including headers from the component. In the future, we can probably do some more cleanup up here. I dont think every component that is linking simplecpp is using headers from simplecpp, but I didnt check the c++ source files and added it as it was being included in the cmake. 

For tinyxml2, I created a target even when we are using an external tinyxml2. So we just need to link the `tinyxml2` and no longer to check if its external and link something else.

The cppcheck-core will still be an object library when `BUILD_SHARED_LIBS=Off` because of the auto-registration of the checks. I realized later that object libraries can now be linked starting in cmake 3.12, which means object libraries can still propagate usage requirements as well. In the future, we could consider making other libraries `OBJECT` as well.

Supporting `BUILD_SHARED_LIBS=On` is beyond the scope of this PR, but I did test this on non-windows and it does work. To support this on windows, will probably just need some small tweaks to support(such as setting `WINDOWS_EXPORT_ALL_SYMBOLS` property or using an object library), but all these could be addressed in the future if the need arises. 

In some places, we were using a couple of source files from another component and rebuilding them again(like filelister.cpp). Instead, I just link the component that builds that source file instead of doing these kind of hacks. If we want to make a subset of source files a seperate component then we could refactor that in the future. I dont do that here since I want to minimize the changes here, but it is something we can address in the future. 

As such, the dmake exe needs the cli lib to be built. So we still build the cli lib even when `BUILD_CLI=Off`, but we wont build a cppcheck executable. I am not sure what the usecase is for `BUILD_CLI=Off` is, but if its just to build less when using the gui, we can work on improving this in the future in order to minimize the changes here.

pfultz2 commented:
@firewave Are you able to review this?

gruenich commented:
* I think most changes are an improvement and should be merged.
* I would drop the non-CMake changes (74f853f8631c1a3e76141d07a1d6e0dba25c1dbc, 91afb23812e1c1bbc642c55213ebaf92931d064f) from this pull request and create one for them alone.
* Several of the commits should be squashed together and get a better commit message with some explanations. (If nobody is willing, I can do so and prove an alternative pull request).

pfultz2 commented:
@firewave I think this can be merged now.

firewave commented on tools/dmake/CMakeLists.txt:
```diff
@@ -1,23 +1,8 @@
 )
-target_include_directories(dmake PRIVATE ${CMAKE_SOURCE_DIR}/cli ${CMAKE_SOURCE_DIR}/lib)
-target_externals_include_directories(dmake PRIVATE ${CMAKE_SOURCE_DIR}/externals/simplecpp)
+target_link_libraries(dmake cppcheck-core cli simplecpp)
```
This will get rid of the matchcompiler hack but I do not like that it now requires the whole libraries to build this.

pfultz2 replied:
If you dont want the whole library then it needs to be separated into a smaller library that can be shared among the components instead. That can be done in a later PR.

firewave commented:
> Still a lot of things in here I have to look at more closely which I unfortunately didn't get around to yet. And there's still way more important things for me to look into then this.

Some clarification. Because of ongoing various personal issues it's often hard for me to concentrate on bigger tasks so I tend to look at some of *my* (not the one of others) older stuff/backlog or low-hanging fruits which usually requires less focused work rather than just some cleaning up. And I also try to avoid things which involve discussions. It also annoys me especially in case of external submissions.

pfultz2 commented:
> Some clarification. Because of ongoing various personal issues it's often hard for me to concentrate on bigger tasks so I tend to look at some of my (not the one of others) older stuff/backlog or low-hanging fruits which usually requires less focused work rather than just some cleaning up. And I also try to avoid things which involve discussions. It also annoys me especially in case of external submissions.

Sure, hope you feel better soon. I dont think you need to be a blocker for this. 

@danmar @chrchr-github @orbitcowboy Can you review and merge this PR since @firewave is not able to right now?

pfultz2 commented on cli/CMakeLists.txt:
```diff
@@ -1,68 +1,37 @@
-if (BUILD_CLI)
```
No, because dmake requires `FileLister` from cli library, and dmake is always built so its always needed. 

The previous hacks just cause a long chain of other problems and hacks. It also breaks the modularity of the components in cppcheck. 

This can be solved by either moving `FileLister` into cppcheck-core or making a seperate library for `FileLister`. I figure this could be handled by a later PR to minimize the changes in this PR.

danmar replied:
> This can be solved by either moving FileLister into cppcheck-core or making a seperate library for FileLister.

how about building dmake if cli is built. I don't know if it has to be built always.

gruenich commented:
> I am not good at cmake. I really hope others can review this more properly.

As I said earlier, this pull request is an improvement to Cppcheck and could be improved itself. It is difficult to grasp the changes and their effect. It could be restructured and split into multiple pull requests. Several commits could be squashed together. If you are interested, I create these pull requests from this one as an alternative proposal.

danmar commented:
> And iz should have been a ticket as well given how serious of a change this is.

:+1: it's never wrong with a ticket. all bigger changes should have tickets.

pfultz2 commented:
> Then we need to actively disallow it and not have some implicitly enabled but probably not usable/working. I do not want any tickets about this or any time at all spent on this in the future. That includes any external contribution which built on top of this because that will leave the support up to us and only cause work down the road - if people need such a feature they have to maintain a fork.

Since this is an open source project, we should be careful about explicitly shutting the door on things that the community can reasonably support or maintain, even if the core contributors don’t personally need that path right now. It’s not a one-man show, if someone wants to keep a config/feature alive and is willing to do the CI/tests/maintenance for it, we should leave that possibility open rather than hard-disabling it up front. They shouldn't need to create a fork for this.

Furthermore, there is no pressure for you to fix tickets either. The invested parties that are interested can handle those issues.

firewave commented:
> Since this is an open source project, we should be careful about explicitly shutting the door on things that the community can reasonably support or maintain, even if the core contributors don’t personally need that path right now.

The problem is that stuff is often just dumped on the project and then you will never see that contributor again. Recently such contribution have not made it in because there had not been any responses on our feedback. Also the CI is more comprehensive so faulty/bad quality contributions will be immediately visible.

Also things which are known not to work should tell you otherwise you think you might have done something wrong or might lead to unexpected results.

> Furthermore, there is no pressure for you to fix tickets either. The invested parties that are interested can handle those issues.

Sometimes I feel like I am the only invested party in some areas. I do the work because I want this to be usable (maintainable is something different) and I feel it still isn't in some areas. At my past job it was a really hard sell to keep it in the CI because of the friction it generated on a regular basis compared to other more integrated (but much inferior or even useless) checkers. Performance was also an issue back then and that was when we were still much fast than now. That's what got me back into the project after years away from it. That unfortunately snowballed badly and got me in the position where I still have too many WIP things.

> I dont even have merge permission to merge those PRs.

I was not aware of that.

> What documentation are you referring to? I didnt make any changes to what we currently document on our readme.

You broke (i.e. incompletely removed) `BUILD_CORE_DLL`. That needs to be put into `releasenotes.txt`. Since we are lacking a ticket we should probably mention the whole rework since this could affect packagers (maybe even worth mentioning regardless).

> But the point is that if we do things properly like in this PR, we dont need all the scenarios tested in CI since it should work automatically.

What? Of course you need to test it. Just because you use a certain pattern/framework stuff won't magically work. If we made a mistake in a target/configuration we never build we will never discover the issue (my point about `BUILD_CORE_DLL` never being build in the CI - this is also which triggered all of this).

pfultz2 commented:
> In `CI-uniyish.yml` the "Generate dependencies" step which is only supposed to generate the Qt files will now build the the whole `lib` and dependencies. So something is wrong there. That needs to be fixed.

I assumed this is because it builds dmake as well, which requires cli and core(as discussed that here: https://github.com/danmar/cppcheck/pull/7658#discussion_r2473185033), however, I dont see `run-dmake` as a dependency here. 

I am not sure how this `add_custom_target(SOURCES)` works as its not really documented to be used this way(the docs says its only for editing files on the IDE). Perhaps because its adding header files which it uses the dependencies of `cppcheck-gui`. Before it was adding the object libraries to the sources so it wasnt a dependency. So it now adds all the dependencies not just the Qt dependencies. I dont think its a huge issue that needs reverting. Perhaps removing the header files might fix it.

pfultz2 commented:
> I agree with @firewave, this pull request should be reverted. The parts we can agree on should be polished and properly described.

I dont see how this could be split up into smaller PRs. What you described below doesnt seem to help split it up. I think reverting is overkill here. 

> Currently it is a mix of different topics, with fixes and revert that should be squashed.

It was squashed.

> The purpose of singe commits is poorly described ("CMake cleanup", "Fix cmake").

That wasnt the only description. It said: "This removes the object libraries and uses normal cmake targets. This helps simplifies a lot of the cmake since we can properly propagate usage requirements.". Unfortunately, I dont have merge access so cant control the commit message there.

> 
>     1. CMake fixes: remve double empty lines,

Removing the double empty lines was probably just my editor. I dont think it needs to be a seperate commit.

> not necessary `foreach` loop

The `foreach` is not necessary due to my changes. It couldn't be a seperate commit.

> 
>     2. C++ changes to fix CI

There was no C++ changes. Those were all fixed in seperate PRs.

> 
>     3. library stuff. Might be to controversial got get it merged, difficult to judge in the current form. The benefit of the change is not clearly described in the commit messages. We might continue the discussion for a cleaned-up pull request with.

That is the meet of the change. It uses cmake targets to handle dependencies instead of using the hacks and copy and pasting we were using before. Its very difficult to maintain and leads(and has led) to bugs. 

This makes the dependencies much simpler to reason about, but it leads to more dependencies to be built, unfortunately due to the hacky nature some of the dependencies are done. So we now can say that A depends on B(ie `A -> B`) instead of A depends on half of B which there isn't a graph notation for that.

In some cases, we depend on the headers of a library and not the sources, but it becomes problematic when you happen to include another header in the same library that will cause a linker error because the library wasnt linked in.

If we want to reduce the dependencies getting built we need to reorganize our dependencies to make them smaller and clearer. Boost had similar but much more severe issues(like circular dependencies) that they worked to resolved when they switched to modular components, but it helped make dependencies much easier to reason about. We can definitely do the same here as we have a lot smaller number of components than boost.
